src/main/java/org/candlepin/controller/util/PromotedContent.java

```diff
@@ -0,0 +1,190 @@
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.controller.util;
+
+import org.candlepin.model.Environment;
+import org.candlepin.model.EnvironmentContent;
+import org.candlepin.model.ProductContent;
+
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+
+/**
+ * A collection of content promoted to environments.
+ *
+ * Stored content from the added environments is de-duplicated based on the environment priority order.
+ */
+public class PromotedContent {
+
+    private static final Logger log = LoggerFactory.getLogger(PromotedContent.class);
+
+    private final Map<String, EnvironmentContent> contents;
+    private final ContentPrefix prefix;
+
+    /**
+     * A constructor
+     *
+     * @param prefix prefixes to be used in path construction
+     */
+    public PromotedContent(ContentPrefix prefix) {
+        this.prefix = Objects.requireNonNull(prefix);
+        this.contents = new HashMap<>();
+    }
+
+    /**
+     * Adds content of the given environment to the promoted content.
+     *
+     * It is important for the environments for come ordered by priority from
+     * highest to lowest as content collected in earlier calls takes precedence.
+     *
+     * @param environment an environment from which to collect content
+     * @return this instance of {@link PromotedContent}
+     */
+    public PromotedContent with(Environment environment) {
+        if (environment == null) {
+            return this;
+        }
+
+        log.debug("Checking for promoted content in environment: {}", environment.getName());
+        for (EnvironmentContent envContent : environment.getEnvironmentContent()) {
+            log.debug("  promoted content: {}", envContent.getContent());
+
+            this.contents.putIfAbsent(envContent.getContent().getId(), envContent);
+        }
+
+        return this;
+    }
+
+    /**
+     * Collects promoted content from all given environments.
+     *
+     * Expects environments to be sorted by priority (from highest to lowest) in
+     * order to deduplicate the promoted content.
+     *
+     * @param environments environments from which to collect promoted content
+     * @return an instance of {@link PromotedContent}
+     */
+    public PromotedContent withAll(List<Environment> environments) {
+        if (environments == null || environments.isEmpty()) {
+            log.debug("No environments to check for promoted content.");
+            return this;
+        }
+
+        log.debug("Consumer has an environment; checking for promoted content.");
```

Comment:
The log line here is out of context and probably unnecessary. Should probably move this back to the cert generation area, or remove it entirely.

Reply:
Done.